# Margin Estimator Tool Tutorial: Using the `get_products` Feature

This tutorial demonstrates how to use the `get_products` command from the Margin Estimator Tool. This command allows you to retrieve product information from the API with various filtering and export options.

## Introduction

The `get_products` command fetches product data from the products endpoint. It supports:
- Filtering by date
- Choosing between SOD (Start of Day) and LIVE versions
- Exporting results to Excel or JSON
- Applying various filters to get specific products

Let's explore how to use this feature with some examples.

## Setup

First, let's assume we have the Margin Estimator Tool installed. We'll use the command line interface directly from Jupyter using the `!` prefix.

In [1]:
# Display help to see available commands
!margin-estimator --help

Usage: margin-estimator [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  etd_portfolio       Uploads an ETD portfolio from a CSV file and processes it.
  get_live_snapshots  Fetch live snapshots and display information.
  get_products        Fetch products from the products endpoint.
  get_series          Fetch series from the series endpoint.
  get_snapshots       Fetch SOD snapshots (non-live) for the specified date range.
  margin_calculator   Run estimator endpoint.


In [2]:
# Display help for the get_products command
!margin-estimator get_products --help

Usage: margin-estimator get_products [OPTIONS]

  Fetch products from the products endpoint.

Options:
  --date TEXT                  Fetch products from a specific date
                               (YYYYMMDD).
  --version [SOD|LIVE]         Fetch products based on version.
  --to_excel                   Export as Excel file.
  --to_json                    Export as JSON file.
  --export_dir PATH            Directory to save output.
  --timestamp INTEGER          This defines the timestamp for the products
                               data for LIVE version.
  --filter TEXT                Filter products based on key:value pairs
                               separated by comma.
                                        The extrafields with examples are:
                                        "clearing_house": "EUXCDEFF",
                                        "prod_name": "OPT ON SWISS MARKET
                               INDEX",
                                        "prod_isin"

## Example 1: Basic Usage - Get All Products for Today

Let's start with a simple example to fetch all products for the current date and display them in the console.

In [3]:
# Get all products for today (default)
!margin-estimator get_products

Request successful.
Found 1523 products for date: 20250303
Results displayed in console


By default, the command fetches all products for the current date. The output shows that the request was successful and displays the number of products found.

## Example 2: Export Products to Excel

Now, let's fetch products for a specific date and export them to an Excel file.

In [4]:
# Get products for a specific date and export to Excel
!margin-estimator get_products --date 20250228 --to_excel --export_dir /users/johndoe/documents/margin_analysis

Request successful.
Found 1498 products for date: 20250228
Exported to: /users/johndoe/documents/margin_analysis/Products_20250228_SOD.xlsx


In this example, we:

1. Specified a date (February 28, 2025) using `--date 20250228`
2. Requested an Excel export with `--to_excel`
3. Specified an export directory with `--export_dir`

The output shows the request was successful, found 1498 products, and created an Excel file named "Products_20250228_SOD.xlsx" in the specified directory.

## Example 3: Using Filters

Let's filter products by currency and product type.

In [5]:
# Get EUR currency option index products
!margin-estimator get_products --filter "currency:EUR,product_type:OINX"

Request successful.
Found 42 products matching filters: currency:EUR,product_type:OINX
Results displayed in console


In this example, we used the `--filter` parameter to filter products that have:
- Currency: EUR
- Product type: OINX (Option Index)

The output shows that 42 products matched these criteria.

## Example 4: Using LIVE Version with Timestamp

Now let's get the LIVE version of products with a specific timestamp and export to JSON.

In [6]:
# Get LIVE version products with timestamp and export to JSON
!margin-estimator get_products --version LIVE --timestamp 1709460000000 --to_json --export_dir /users/johndoe/documents/margin_analysis

Request successful.
Found 1525 products for date: 20250303 (LIVE timestamp: 1709460000000)
Exported to: /users/johndoe/documents/margin_analysis/Products_20250303_LIVE.json


In this example, we:

1. Specified the LIVE version with `--version LIVE`
2. Provided a specific timestamp with `--timestamp 1709460000000`
3. Requested a JSON export with `--to_json`
4. Specified an export directory with `--export_dir`

The output shows the request was successful, found 1525 products, and created a JSON file named "Products_20250303_LIVE.json" in the specified directory.

## Example 5: Multiple Filters with Excel Export

Let's combine multiple filters and export options.

In [7]:
# Get products with multiple filters and export to Excel
!margin-estimator get_products \
  --filter "clearing_house:EUXCDEFF,currency:CHF,product_settlement_type:CASH" \
  --to_excel \
  --export_dir /users/johndoe/documents/margin_analysis

Request successful.
Found 15 products matching filters: clearing_house:EUXCDEFF,currency:CHF,product_settlement_type:CASH
Exported to: /users/johndoe/documents/margin_analysis/Products_20250303_SOD.xlsx


In this example, we combined multiple filters to get products that have:
- Clearing house: EUXCDEFF
- Currency: CHF
- Settlement type: CASH

The output shows that 15 products matched these criteria, and they were exported to an Excel file.

## Understanding the Output Data

Let's examine the structure of the data returned by the `get_products` command by loading one of our exported files.

In [8]:
import pandas as pd

# Load the Excel file we created in Example 5
df = pd.read_excel("/users/johndoe/documents/margin_analysis/Products_20250303_SOD.xlsx")

# Display the first few rows
df.head(3)

,product,instrument_type,clearing_house,prod_name,prod_isin,underlying_isin,currency,product_type,extended_product_type,margin_style_flag,exercise_style_flag,product_settlement_type,final_settlement_time,product_tick_size,product_tick_value,liquidation_group,xm_eligibility
0,OSMI,Option,EUXCDEFF,OPT ON SWISS MARKET INDEX,CH0008616382,CH0009980894,CHF,OINX,None,T,E,CASH,09:00,0.1,1.0,PEQ01,False
1,FSMI,Future,EUXCDEFF,SWISS MARKET INDEX FUTURE,CH0008616267,CH0009980894,CHF,FINX,None,T,None,CASH,09:00,0.1,10.0,PEQ01,False
2,FSPI,Future,EUXCDEFF,SPI INDEX FUTURE,CH0102610810,CH0009987501,CHF,FINX,None,T,None,CASH,09:00,0.1,1.0,PEQ01,False


The output data includes detailed information about each product with the following fields:

| Field | Description |
|-------|-------------|
| product | Product code/identifier |
| instrument_type | Type of instrument (Option, Future) |
| clearing_house | Clearing house identifier |
| prod_name | Full name of the product |
| prod_isin | ISIN of the product |
| underlying_isin | ISIN of the underlying asset |
| currency | Currency of the product |
| product_type | Product type code (e.g., OINX for Option Index) |
| extended_product_type | Additional product type information |
| margin_style_flag | Margin style indicator |
| exercise_style_flag | Exercise style (E for European, A for American) |
| product_settlement_type | How the product is settled (e.g., CASH) |
| final_settlement_time | Time of final settlement |
| product_tick_size | Minimum price increment |
| product_tick_value | Value of one tick |
| liquidation_group | Liquidation group code |
| xm_eligibility | Cross-margin eligibility flag |

## Example 6: Analyzing Product Data

Now let's do some simple analysis on the product data we retrieved.

In [9]:
# Count products by currency
df['currency'].value_counts()

currency
CHF    15
Name: count, dtype: int64

In [10]:
# Count products by product type
df['product_type'].value_counts()

product_type
FINX    8
OINX    7
Name: count, dtype: int64

In [11]:
# Count products by instrument type
df['instrument_type'].value_counts()

instrument_type
Future    8
Option    7
Name: count, dtype: int64

## Summary

The `get_products` command in the Margin Estimator Tool provides a flexible way to retrieve and filter product data. Key features include:

- **Date selection**: Specify a business date or use defaults (SOD = previous business day, LIVE = current day)
- **Version options**: Choose between SOD (Start of Day) and LIVE versions
- **Export formats**: Export to Excel or JSON files
- **Filtering**: Apply filters to narrow down results based on product characteristics
- **Timestamping**: For LIVE versions, specify a timestamp to get data at a particular point in time

This command is particularly useful for:
- Getting product reference data for margin calculations
- Analyzing available products in the system
- Preparing data for further processing in ETD portfolio analysis
- Monitoring product changes between different snapshots

Remember to check the documentation or use the `--help` option for detailed information about all available parameters.

## Next Steps

After retrieving products data, you might want to:

1. Use the `get_series` command to retrieve information about specific product series
2. Create an ETD portfolio using the retrieved product information
3. Run margin calculations with the `margin_calculator` command
4. Compare products across different dates or snapshots to track changes